07.ls_object/file_display.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+require 'optparse'
+
+class FileDisplay
+  NUMBER_OF_COLUMNS = 3
+
+  def initialize
+    @options = ARGV.getopts('alr').transform_keys(&:to_sym)
+    @file_stats = fetch_file_stats
+  end
+
+  def select_format
+    @options[:l] ? long_format : short_format
+  end
+
+  private
+
+  def fetch_file_stats
+    flags = @options[:a] ? File::FNM_DOTMATCH : 0
+    file_names = Dir.glob('*', flags)
+    sorted_file_names = @options[:r] ? file_names.reverse : file_names
+    sorted_file_names.map { |file_name| FileStat.new(file_name) }
+  end
+
+  def long_format
+    blocks_total = @file_stats.map(&:blocks).sum
+    puts "total #{blocks_total}"
+
+    @file_stats.each { |file_stat| puts format_file_stat(file_stat) }
+  end
+
+  def format_file_stat(file_stat)
+    stat_map = build_stat_map(file_stat)
+    max_size_map = build_max_size_map(@file_stats)
+
+    format([
+      '%<file_mode>s',
+      "%<nlink>#{max_size_map[:nlink] + 1}s",
+      "%<user_name>-#{max_size_map[:user_name] + 1}s",
+      "%<group_name>-#{max_size_map[:group_name] + 1}s",
+      "%<size>#{max_size_map[:size]}s",
+      '%<mtime>2s',
+      "%<name>s\n"
+    ].join(' '), stat_map)
+  end
+
+  def build_stat_map(file_stat)
+    {
+      file_mode: file_stat.file_mode,
+      nlink: file_stat.nlink,
+      user_name: file_stat.user.name,
```

Comment:
```suggestion
      user_name: file_stat.user_name,
```

デメテルの法則を考えると、こうなってる方が良いかも？

Reply:
承知しました。

デメテルの法則とは、オブジェクトのメソッド呼び出しが連らなるのを防ごうという法則なんですね。

確かに、`file_stat.user.name`を`file_stat.user_name`にすることで、`user_name`の処理に変更があるときには、`user_name`の

【参考記事】
[お前のデメテルの法則は間違っている ～getter/setterの必要性～](https://tech-blog.rakus.co.jp/entry/20200701/programming)